### PR TITLE
[DO-927] Export pep to xlsx

### PR DESCRIPTION
**Pull request by Iva-khar**

### CODE REVIEW

File: business_register/views/pep_views.py [hunk 1]
```diff
@@ -43,3 +46,30 @@ def retrieve_by_source_id(self, request, pk):
         pep = get_object_or_404(self.get_queryset(), source_id=pk)
         serializer = self.get_serializer(pep)
         return Response(serializer.data)
+
+    @action(detail=False, url_path='xlsx')
+    def export_to_xlsx(self, request):
+        before_date = datetime.strptime(request.GET['updated_at_before'], '%Y-%m-%d')
+        after_date = datetime.strptime(request.GET['updated_at_after'], '%Y-%m-%d')
```
Comment (roman-tiukh) [inline business_register/views/pep_views.py:53]: what hapen if get-parameters is wrong or not exists?

File: business_register/views/pep_views.py [hunk 2]
```diff
@@ -43,3 +46,30 @@ def retrieve_by_source_id(self, request, pk):
...
+            'Full Name (english transcription)': ['fullname_transcriptions_eng', 36],
+            'Status': ['is_pep', 10],
+            'PEP Type': ['pep_type', 10],
+            'Date of Birth': ['date_of_birth', 19],
+            'Created Date': ['created_at', 19],
+            'Updated Date': ['updated_at', 19],
+            'Last Job Title': ['last_job_title', 20],
+            'Last Employer': ['last_employer', 20]
+        }
+        worksheet_title = 'PEP'
```
Comment (roman-tiukh) [inline business_register/views/pep_views.py:72]: why you create this variable?

File: business_register/views/pep_views.py (hunk 1 above)
Comment (roman-tiukh) [reply business_register/views/pep_views.py:53]: If there are no parameters, then just return the data no older than 1 month

File: business_register/views/pep_views.py [hunk 3]
```diff
@@ -43,3 +46,30 @@ def retrieve_by_source_id(self, request, pk):
...
+        export_file_path = ExportToXlsx().export(queryset, export_dict, worksheet_title)
+        if export_file_path:
+            return HttpResponse(export_file_path, content_type="text/plain")
```
Comment (roman-tiukh) [inline business_register/views/pep_views.py:75]: else??

Comment (roman-tiukh) [inline business_register/views/pep_views.py:75]: Use `Response` from rest framework and return json

File: business_register/views/pep_views.py [hunk 4]
```diff
@@ -43,3 +46,30 @@ def retrieve_by_source_id(self, request, pk):
...
+    @action(detail=False, url_path='xlsx')
+    def export_to_xlsx(self, request):
+        before_date = datetime.strptime(request.GET['updated_at_before'], '%Y-%m-%d')
+        after_date = datetime.strptime(request.GET['updated_at_after'], '%Y-%m-%d')
+        if (before_date and after_date) and timedelta(days=0) < before_date - after_date <= timedelta(days=31):
+            queryset = self.filter_queryset(self.get_queryset())
+        else:
+            return HttpResponse('Use "updated_at_after" & "updated_at_before" GET parameters. '
+                                'The time period for data exported in .xlsx cannot exceed 31 days.',
+                                content_type="text/plain")
```
Comment (roman-tiukh) [inline business_register/views/pep_views.py:59]: no need this validation

File: business_register/views/pep_views.py [hunk 5]
```diff
@@ -43,3 +46,30 @@ def retrieve_by_source_id(self, request, pk):
...
+            'Status': ['is_pep', 10],
+            'PEP Type': ['pep_type', 10],
+            'Date of Birth': ['date_of_birth', 19],
+            'Created Date': ['created_at', 19],
+            'Updated Date': ['updated_at', 19],
+            'Last Job Title': ['last_job_title', 20],
+            'Last Employer': ['last_employer', 20]
+        }
+        worksheet_title = 'PEP'
+        export_file_path = ExportToXlsx().export(queryset, export_dict, worksheet_title)
```
Comment (roman-tiukh) [inline business_register/views/pep_views.py:73]: ... 
With such success you could not make a class, but make a function. You're not taking advantage of classes and OOP.
ExportToXlsx class is useless, only export method works.
```
ExportToXlsx().export(queryset, export_dict, worksheet_title)
```

File: data_ocean/export.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+                    cell.value = sell_value
+                except:
+                    cell.value = repr(sell_value)
+        export_file_name = worksheet_title + '_{0}.xlsx'.format(datetime.now().strftime('%Y-%m-%d_%H-%M-%S'))
+        with NamedTemporaryFile() as temp_file:
+            workbook.save(temp_file.name)
+            temp_file.seek(0)
+            data = open(temp_file.name, 'rb')
+        s3 = boto3.resource('s3')
+        s3.Bucket('pep-xlsx').put_object(Key=export_file_name, Body=data, ACL='public-read')
```
Comment (roman-tiukh) [inline data_ocean/export.py:43]: where is credentials?
They should located in settinds_local...
Is it possible?

File: data_ocean/export.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+                sell_value = getattr(record, column_properties[0])
+                row_num += 1
+                cell = worksheet.cell(row=row_num, column=col_num)
+                cell.alignment = Alignment(vertical='top', wrap_text=True)
+                try:
+                    cell.value = sell_value
+                except:
+                    cell.value = repr(sell_value)
+        export_file_name = worksheet_title + '_{0}.xlsx'.format(datetime.now().strftime('%Y-%m-%d_%H-%M-%S'))
+        with NamedTemporaryFile() as temp_file:
```
Comment (roman-tiukh) [inline data_ocean/export.py:38]: You dont need NamedTemporaryFile, use `io.BytesIO` instead.

File: data_ocean/export.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+            cell.fill = PatternFill(bgColor='0033CCCC', fill_type="solid")
+            for record in queryset:
+                sell_value = getattr(record, column_properties[0])
+                row_num += 1
+                cell = worksheet.cell(row=row_num, column=col_num)
+                cell.alignment = Alignment(vertical='top', wrap_text=True)
+                try:
+                    cell.value = sell_value
+                except:
+                    cell.value = repr(sell_value)
```
Comment (roman-tiukh) [inline data_ocean/export.py:36]: what is reason for catching **ALL** exceptions?

File: data_ocean/export.py [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+        for col_num, (column_title, column_properties) in enumerate(export_dict.items(), 1):
+            row_num = 1
+            cell = worksheet.cell(row=row_num, column=col_num)
+            cell.value = column_title
+            worksheet.column_dimensions[get_column_letter(col_num)].width = column_properties[1]
+            cell.alignment = Alignment(horizontal='center', vertical='center')
+            cell.font = fonts.Font(b=True)
+            cell.fill = PatternFill(bgColor='0033CCCC', fill_type="solid")
+            for record in queryset:
+                sell_value = getattr(record, column_properties[0])
```
Comment (roman-tiukh) [inline data_ocean/export.py:29]: sell value?
image

File: data_ocean/export.py [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+                except:
+                    cell.value = repr(sell_value)
+        export_file_name = worksheet_title + '_{0}.xlsx'.format(datetime.now().strftime('%Y-%m-%d_%H-%M-%S'))
+        with NamedTemporaryFile() as temp_file:
+            workbook.save(temp_file.name)
+            temp_file.seek(0)
+            data = open(temp_file.name, 'rb')
+        s3 = boto3.resource('s3')
+        s3.Bucket('pep-xlsx').put_object(Key=export_file_name, Body=data, ACL='public-read')
+        return settings.PEP_EXPORT_FOLDER_URL + export_file_name
```
Comment (roman-tiukh) [inline data_ocean/export.py:44]: is that full(absolute) link to file in S3?

Comment (roman-tiukh) [reply data_ocean/export.py:44]: what is PEP_EXPORT_FOLDER_URL?

File: README.md [hunk 11]
```diff
@@ -296,3 +296,49 @@ Debugging in VSCode
 -----------------------------------------------------------------------------------------------------
 Official documentation:
 - https://code.visualstudio.com/docs/editor/debugging
+
+-----------------------------------------------------------------------------------------------------
+Export files to AWS S3
+-----------------------------------------------------------------------------------------------------
+Source: https://boto3.amazonaws.com/
```
Comment (roman-tiukh) [inline README.md:303]: image

**Review by roman-tiukh — CHANGES_REQUESTED** [review]

File: data_ocean/export.py (hunk 8 above)
Comment (roman-tiukh) [reply data_ocean/export.py:36]: `repr`? Why not `str`?

**Review by roman-tiukh — CHANGES_REQUESTED** [review]
Generation and upload must be processed by Celery as background task.

File: business_register/views/pep_views.py (hunk 2 above)
Comment (Iva-khar) [reply business_register/views/pep_views.py:72]: For sending in `export()` -> for exel sheet naming & file naming

File: business_register/views/pep_views.py (hunk 3 above)
Comment (Iva-khar) [reply business_register/views/pep_views.py:75]: `if` deleted

File: business_register/views/pep_views.py (hunk 1 above)
Comment (Iva-khar) [reply business_register/views/pep_views.py:53]: The get-parameters validation have been refactored

File: business_register/views/pep_views.py (hunk 2 above)
Comment (Iva-khar) [reply business_register/views/pep_views.py:72]: Rewrited with literal usage

File: business_register/views/pep_views.py (hunk 3 above)
Comment (Iva-khar) [reply business_register/views/pep_views.py:75]: done

File: business_register/views/pep_views.py (hunk 4 above)
Comment (Iva-khar) [reply business_register/views/pep_views.py:59]: The get-parameters validation have been refactored

File: business_register/views/pep_views.py (hunk 5 above)
Comment (Iva-khar) [reply business_register/views/pep_views.py:73]: Will refactor after the new export views appears..

File: data_ocean/export.py (hunk 6 above)
Comment (Iva-khar) [reply data_ocean/export.py:43]: done

File: data_ocean/export.py (hunk 7 above)
Comment (Iva-khar) [reply data_ocean/export.py:38]: done

File: data_ocean/export.py (hunk 8 above)
Comment (Iva-khar) [reply data_ocean/export.py:36]: (try-exept) - deleted

File: data_ocean/export.py (hunk 9 above)
Comment (Iva-khar) [reply data_ocean/export.py:29]: fixed

File: README.md (hunk 11 above)
Comment (Iva-khar) [reply README.md:303]: Readmy for expord has deleted

**Comment by Iva-khar** [comment]
> Generation and upload must be processed by Celery as background task.

done

**Comment by Iva-khar** [comment]
accidentally deleted repited comment "repr? Why not str?"

**Review by roman-tiukh — CHANGES_REQUESTED** [review]
image

File: business_register/views/pep_views.py [hunk 12]
```diff
@@ -43,3 +44,34 @@ def retrieve_by_source_id(self, request, pk):
...
+            'Date of Birth': ['date_of_birth', 19],
+            'Created Date': ['created_at', 19],
+            'Updated Date': ['updated_at', 19],
+            'Last Job Title': ['last_job_title', 20],
+            'Last Employer': ['last_employer', 20]
+        }
+        from business_register.tasks import export_to_s3
+        export_to_s3.delay(query, export_dict, 'Pep')
+        data = {"detail": "Generation of .xlsx file has begin. Expect an email with downloading link."}
+        return Response(data, status=200)
```
[the remark this answers is not in the document]
Comment (Iva-khar) [reply business_register/views/pep_views.py:77]: done

File: business_register/tasks.py [hunk 13]
```diff
@@ -60,3 +60,14 @@ def update_uk_company():
...
+
+@shared_task
+def export_to_s3(query, export_dict, model_name):
+    print('***************************')
+    print(' Export ' + model_name + ' to S3')
+    print('***************************')
+
+    from data_ocean.export import ExportToXlsx
+
+    return ExportToXlsx().export(query, export_dict, model_name)
```
[the remark this answers is not in the document]
Comment (Iva-khar) [reply business_register/tasks.py]: done

File: business_register/views/pep_views.py [hunk 14]
```diff
@@ -43,3 +44,34 @@ def retrieve_by_source_id(self, request, pk):
...
+
+    @action(detail=False, url_path='xlsx')
+    def export_to_xlsx(self, request):
+        filterset = DjangoFilterBackend().get_filterset(request, self.get_queryset(), self)
+        form = filterset.form
+        form.full_clean()
+        cleaned_data = form.cleaned_data
+        if 'updated_at' in cleaned_data and timedelta(days=0) < \
+                cleaned_data['updated_at'].stop - cleaned_data['updated_at'].start <= timedelta(days=30):
+            query = self.filter_queryset(self.get_queryset()).query.__str__()
```
[the remark this answers is not in the document]
Comment (Iva-khar) [reply business_register/views/pep_views.py]: ommited

File: business_register/views/pep_views.py [hunk 15]
```diff
@@ -43,3 +44,34 @@ def retrieve_by_source_id(self, request, pk):
...
+        else:
+            cleaned_data['updated_at'] = slice(datetime.now() - timedelta(days=30), datetime.now(), None)
+            filterset = PepFilterSet(cleaned_data, self.get_queryset())
+            queryset = filterset.qs
+            query = SearchFilter().filter_queryset(request, queryset, self).query.__str__()
```
[the remark this answers is not in the document]
Comment (Iva-khar) [reply business_register/views/pep_views.py]: ommmited

File: business_register/emails.py [hunk 16]
```diff
@@ -13,3 +14,16 @@ def send_new_company_type_message(company_type: CompanyType):
             'company_type': company_type,
         }
     )
+
+
+def send_export_url_file_path_message(user_id: int, link: str):
+    user = DataOceanUser(id=user_id)
```
Comment (roman-tiukh) [inline business_register/emails.py:20]: ...

File: business_register/emails.py [hunk 17]
```diff
@@ -13,3 +14,16 @@ def send_new_company_type_message(company_type: CompanyType):
...
+    send_template_mail(
+        to=[user.email],
+        subject='Generation of .xlsx export file completed',
```
Comment (roman-tiukh) [inline business_register/emails.py:23]: At end it must be localized

File: data_ocean/export.py [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+from business_register.emails import send_export_url_file_path_message
+from users.models import DataOceanUser
+
+
+class ExportToXlsx:
+
+    @staticmethod
+    def export(params, export_dict, model_name, user_id):
+        queryset = apps.get_model('business_register', model_name).objects.all()
+        export_filterset_class = getattr(filters, model_name + 'ExportFilterSet')
```
Comment (roman-tiukh) [inline data_ocean/export.py:22]: i think its wrong way...

Comment (Iva-khar) [reply data_ocean/export.py:22]: refactored

File: business_register/emails.py (hunk 16 above)
Comment (Iva-khar) [reply business_register/emails.py:20]: fixed

File: business_register/emails.py (hunk 17 above)
Comment (Iva-khar) [reply business_register/emails.py]: In another PR.

File: business_register/forms.py [hunk 19]
```diff
@@ -0,0 +1,19 @@
...
+
+
+class PepExportForm(forms.Form):
+
+    def clean(self):
+        cleaned_data = super().clean()
+        updated_at = cleaned_data.get('updated_at', None)
+        if updated_at is None or updated_at.start is None or updated_at.stop is None:
+            raise forms.ValidationError({
+                'updated_at': ['Period for updated_at is not provided.']
```
Comment (roman-tiukh) [inline business_register/forms.py:12]: localization

File: business_register/forms.py [hunk 20]
```diff
@@ -0,0 +1,19 @@
...
+            })
+        elif not timedelta(days=0) < updated_at.stop - updated_at.start <= timedelta(days=30):
+            raise forms.ValidationError({
+                'updated_at': ['Period for updated_at not matches restrictions.']
```
Comment (roman-tiukh) [inline business_register/forms.py:16]: localization

**Review by roman-tiukh — CHANGES_REQUESTED** [review]
no generated messages for localization

File: data_ocean/locale/uk/LC_MESSAGES/django.po [hunk 21]
```diff
@@ -21,104 +21,120 @@ msgstr ""
...
 msgstr "Всі права захищені"
 
-#: data_ocean/templates/data_ocean/emails/hi_there.html:3
+#: templates/data_ocean/emails/export_to_xlsx_completed.html:8
+msgid "We have generated the .xlsx file for your request. Now You may"
+msgstr "Файл формату .xlsx за Вашим запитом сгенеровано."
+
+#: templates/data_ocean/emails/export_to_xlsx_completed.html:12
+msgid "download it"
+msgstr "Скачати файл"
```
Comment (roman-tiukh) [inline data_ocean/locale/uk/LC_MESSAGES/django.po:136]: diwnload it -> скачати файл?

File: data_ocean/locale/uk/LC_MESSAGES/django.po [hunk 22]
```diff
@@ -21,104 +21,120 @@ msgstr ""
...
 
-#: data_ocean/templates/data_ocean/emails/_base.html:52
+#: templates/data_ocean/emails/_base.html:52
 msgid "All rights reserved"
 msgstr "Всі права захищені"
 
-#: data_ocean/templates/data_ocean/emails/hi_there.html:3
+#: templates/data_ocean/emails/export_to_xlsx_completed.html:8
+msgid "We have generated the .xlsx file for your request. Now You may"
+msgstr "Файл формату .xlsx за Вашим запитом сгенеровано."
```
Comment (roman-tiukh) [inline data_ocean/locale/uk/LC_MESSAGES/django.po:132]: Now You may?

File: data_ocean/locale/uk/LC_MESSAGES/django.po [hunk 23]
```diff
@@ -21,104 +21,120 @@ msgstr ""
...
 msgid "Remove"
 msgstr "Видалити"
 
-#: data_ocean/templates/data_ocean/emails/_base.html:24
+#: templates/data_ocean/emails/_base.html:24
 msgid "Regards"
 msgstr "З повагою"
 
-#: data_ocean/templates/data_ocean/emails/_base.html:48
+#: templates/data_ocean/emails/_base.html:48
```
Comment (roman-tiukh) [inline data_ocean/locale/uk/LC_MESSAGES/django.po:122]: data_ocean markers was deleted?

File: data_ocean/export.py [hunk 24]
```diff
@@ -0,0 +1,68 @@
...
+            aws_secret_access_key=settings.AWS_SECRET_ACCESS_KEY,
+            config=config
+        )
+        s3.Bucket('pep-xlsx').put_object(Key=export_file_name, Body=data, ACL='public-read')
+        export_url = settings.PEP_EXPORT_FOLDER_URL + export_file_name
+        user = DataOceanUser.objects.get(id=user_id)
+        with translation.override(user.language):
+            user.notify(
+                _('Generation of .xlsx file has ended. You may download the file by link:'),
+                export_url
```
Comment (roman-tiukh) [inline data_ocean/export.py:66]: `export_url` it's not a continue for message. The link is defined for clicking on the message

File: requirements.txt [hunk 25]
```diff
@@ -40,18 +42,22 @@ future==0.18.2
 html5lib==1.1
 humanize==2.6.0
 idna==2.9
+importlib-metadata==3.7.3
```
Comment (roman-tiukh) [inline requirements.txt:45]: its for boto3?

File: data_ocean/locale/uk/LC_MESSAGES/django.po (hunk 21 above)
Comment (Iva-khar) [reply data_ocean/locale/uk/LC_MESSAGES/django.po:136]: changed

File: data_ocean/locale/uk/LC_MESSAGES/django.po (hunk 22 above)
Comment (Iva-khar) [reply data_ocean/locale/uk/LC_MESSAGES/django.po:132]: changed

File: data_ocean/locale/uk/LC_MESSAGES/django.po (hunk 23 above)
Comment (Iva-khar) [reply data_ocean/locale/uk/LC_MESSAGES/django.po:122]: Yes, by the `makemessages -a`

File: data_ocean/export.py (hunk 24 above)
Comment (Iva-khar) [reply data_ocean/export.py:66]: changed

File: requirements.txt (hunk 25 above)
Comment (Iva-khar) [reply requirements.txt:45]: for boto3 or for openxlsx

File: requirements.txt [hunk 26]
```diff
@@ -85,6 +92,7 @@ starkbank-ecdsa==1.1.0
 tinycss2==1.1.0
 toml==0.10.1
 tornado==6.0.4
+typing-extensions==3.7.4.3
```
Comment (roman-tiukh) [inline requirements.txt:95]: unused

File: requirements.txt [hunk 27]
```diff
@@ -93,3 +101,4 @@ webencodings==0.5.1
 wrapt==1.12.1
 xmltodict==0.12.0
 zeep==3.4.0
+zipp==3.4.1
```
Comment (roman-tiukh) [inline requirements.txt:104]: unused

File: requirements.txt [hunk 28]
```diff
@@ -40,18 +42,22 @@ future==0.18.2
 html5lib==1.1
 humanize==2.6.0
 idna==2.9
+importlib-metadata==3.7.3
 inflection==0.5.0
 isodate==0.6.0
 isort==4.3.21
 itypes==1.2.0
+jdcal==1.4.1
```
Comment (roman-tiukh) [inline requirements.txt:50]: unused

File: requirements.txt (hunk 25 above)
Comment (roman-tiukh) [inline requirements.txt:45]: unused

**Review by roman-tiukh — CHANGES_REQUESTED** [review]

File: requirements.txt [hunk 29]
```diff
@@ -40,18 +42,22 @@ future==0.18.2
...
 Jinja2==2.11.2
+jmespath==0.10.0
 kombu==4.6.10
 lazy-object-proxy==1.4.3
 lxml==4.6.2
 MarkupSafe==1.1.1
 mccabe==0.6.1
 num2words==0.5.10
 oauthlib==3.1.0
+openpyxl==3.0.5
```
[the remark this answers is not in the document]
Comment (Iva-khar) [reply requirements.txt]: added

**Review by roman-tiukh — APPROVED** [review]

**Pull request merged**